patches/0004-Output-to-hashed-mappings.patch

```diff
@@ -0,0 +1,1622 @@
+ import java.util.HashSet;
+ import java.util.List;
++import java.util.Map;
+ import java.util.Set;
+ 
+ import com.google.common.base.Preconditions;
+@@ -146,42 +148,57 @@ public class TransitiveAccessWidenerJarProcessor implements JarProcessor {
+ 
+ 	private AccessWidener createAccessWidener() {
+ 		AccessWidener accessWidener = new AccessWidener();
+-		// For other mods, only consider transitive AWs and remap from intermediary->named
+-		TinyRemapper tinyRemapper = createTinyRemapper();
++		// For other mods, only consider transitive AWs and remap from hashed->named or intermediary->named
++		TinyRemapper hashedRemapper = createTinyRemapper(MappingsNamespace.HASHED);
++		TinyRemapper intermediaryRemapper = createTinyRemapper(MappingsNamespace.INTERMEDIARY);
+ 
+ 		try {
+-			AccessWidenerRemapper remappingVisitor = new AccessWidenerRemapper(
++			AccessWidenerRemapper hashedRemappingVisitor = new AccessWidenerRemapper(
+ 					accessWidener,
+-					tinyRemapper.getEnvironment().getRemapper(),
++					intermediaryRemapper.getEnvironment().getRemapper(),
++					MappingsNamespace.HASHED.toString(),
++					MappingsNamespace.NAMED.toString()
++			);
++			AccessWidenerReader hashedTransitiveReader = new AccessWidenerReader(new TransitiveOnlyFilter(hashedRemappingVisitor));
++
++			AccessWidenerRemapper intermediaryRemappingVisitor = new AccessWidenerRemapper(
++					accessWidener,
++					intermediaryRemapper.getEnvironment().getRemapper(),
+ 					MappingsNamespace.INTERMEDIARY.toString(),
+ 					MappingsNamespace.NAMED.toString()
+ 			);
+-			AccessWidenerReader transitiveReader = new AccessWidenerReader(new TransitiveOnlyFilter(remappingVisitor));
++			AccessWidenerReader intermediaryTransitiveReader = new AccessWidenerReader(new TransitiveOnlyFilter(intermediaryRemappingVisitor));
++
++			final Map<String, AccessWidenerReader> awReaderMap = new HashMap<>();
++			awReaderMap.put(MappingsNamespace.HASHED.toString(), hashedTransitiveReader);
++			awReaderMap.put(MappingsNamespace.INTERMEDIARY.toString(), intermediaryTransitiveReader);
+ 
+ 			for (AccessWidenerFile accessWidenerFile : transitiveAccessWideners) {
+ 				project.getLogger().info("Reading transitive access widener from {}", accessWidenerFile.modId());
+-				transitiveReader.read(accessWidenerFile.content());
++				String awNamespace = AccessWidenerReader.readHeader(accessWidenerFile.content()).getNamespace();
++				awReaderMap.get(awNamespace).read(accessWidenerFile.content());
+ 			}
+ 		} finally {
+-			tinyRemapper.finish();
++			hashedRemapper.finish();
++			intermediaryRemapper.finish();
+ 		}
+ 
+ 		return accessWidener;
+ 	}
+ 
+-	private TinyRemapper createTinyRemapper() {
++	private TinyRemapper createTinyRemapper(MappingsNamespace intermediateMappings) {
+ 		try {
+-			TinyRemapper tinyRemapper = TinyRemapperHelper.getTinyRemapper(project, "intermediary", "named");
++			TinyRemapper tinyRemapper = TinyRemapperHelper.getTinyRemapper(project, "hashed", "named");
```

Comment:
Shouldn't the namespace depend on which intermediateMappings you provided in the argument?

Reply:
Oh, yes, I might have left that in from debugging